### ISSUE TITLE
[Q] how is determined whether savegames can be used

### ISSUE THREAD

**Issue by rofl0r** [adventuregamestudio/ags#1168]
i noticed that after recompiling some source files and repacking a game with agsutils, the savegames were no longer picked up. why is that, and what determines whether savegames are "compatible", and is it possible to workaround those checks somehow ?

**Comment by ivan-mogilko** [comment]
Could you clarify what exactly happened, was there an error message?

**Comment by rofl0r** [comment]
with master branch, the list was empty, with SDL2 branch i see "INVALID SLOT 1" for my old savegame.

**Comment by ivan-mogilko** [comment]
I think this supposed to happen only if save has unsupported format, although I cannot explain difference between master and sdl2 branch.
It's also possible that master branch has got a new bug.
Which branch (or tag etc) did you use to create these saves?

**Comment by rofl0r** [comment]
i used master from like 2 days ago, with about 5 custom patches backported, however they're all related to debug info only when --fps switch is added, and SDL2 branch from yesterday respectively.

attaching my repacked larry2 game with saves (recompiled all room sources with ascc to remove the string corruption).
[larry2_repack.zip](https://github.com/adventuregamestudio/ags/files/5811535/larry2_repack.zip)

**Comment by ivan-mogilko** [comment]
Pushed a fix into master, the search path was created wrong when filling save list.

It currently finds two saves: one of them is indicated as "INVALID SLOT".

**Comment by rofl0r** [comment]
> It currently finds two saves: one of them is indicated as "INVALID SLOT".

right, i didn't know which one was the good and which one the bad save.

> Pushed a fix into master

cool, thanks!

btw regarding the issue with CTRL-E not being picked up with --fps in the intro of 10 minutes - should i open a new issue for that ?

**Comment by rofl0r** [comment]
mentioned commit is https://github.com/adventuregamestudio/ags/commit/7a8aa7077d084266295802dc1805d56abccacbc1 btw

**Comment by ivan-mogilko** [comment]
Sorry, there seem to be more problems. I think what is happening is that game sees "agssave.001_" when scanning for save files, but when trying to load or preload one it uses wrong assumption (default name pattern) and forms "agssave.001" name instead.

**Comment by ivan-mogilko** [comment]
> btw regarding the issue with CTRL-E not being picked up with --fps in the intro of 10 minutes - should i open a new issue for that ?

Yes, I think this may be fixed.

**Comment by ivan-mogilko** [comment]
Pushed few more fixes, this commit fe1617c7bae242a0ee90723f41c1e3cfb7a8ae47 in particular corrects search pattern to avoid including files with similar but not exact match. Now files like "agssave.001_" will be simply ignored.

**Comment by rofl0r** [comment]
i wonder how agssave.001_ came to exist in the first place ?

**Comment by ivan-mogilko** [comment]
I don't know, it was in the archive you sent me, and I assumed you just made a backup of older save?

**Comment by ivan-mogilko** [comment]
By the way, I must mention, this game *seem* to have the old script bug from default game template, where it would incorrectly overwrite existing slot instead of creating a new one if one of the early slot numbers is missing.

Because earlier engine was adding "agssave.001_" into the list this problem was averted, but now when it does not the game practically stops creating new saves. To fix this you have to add missing "agssave.001" file into saves dir.

**Comment by rofl0r** [comment]
> If you did not, then something strange happened.

indeed, i didn't, so the engine must be to blame.

> old script bug from default game template

do you have some reference for this? is this an issue with the example game shipped with old ags versions ?

> To fix this you have to add missing "agssave.001" file into saves dir.

can i just create an empty file of that name ?

**Comment by ivan-mogilko** [comment]
Well, I could not figure out a reason why an underscore may be added to the file name, so unless there's a way to reproduce this I would not know how to address it.

> do you have some reference for this? is this an issue with the example game shipped with old ags versions ?

Not example game, but the template, used as a starting set of files for a game project.
We actually have an old commit that fixed it in template called "Default Game":
https://github.com/adventuregamestudio/ags-template-source/commit/b0928130201d4f81d8c61b2421179916ba8e787a

> can i just create an empty file of that name ?

Yes, I think, except it will be displayed as "invalid slot".

**Comment by rofl0r** [comment]
> We actually have an old commit that fixed it in template called "Default Game":
adventuregamestudio/ags-template-source@b092813

oh wow, i didn't know of this repo. this is super-helpful for reversing old games

**Comment by rofl0r** [comment]
> this is super-helpful for reversing old games

on this topic... i noticed old studio versions ship these as .agt files. are you aware of any tool to extract them, or some code in ags repo (current or past) that can open them ?

**Comment by ivan-mogilko** [comment]
We still do use these, and these are same format as game package iirc, only include project source rather than compiled game.

**Comment by rofl0r** [comment]
on this topic, i created a github org and website https://ags-archives.github.io/
currently it has all old AGS (not AC) engine versions (extracted) and extracted ags-help.chm for each version. the latter proofed really useful and patching agsdefns.sh to support compiling older source code with agscc, the former contains also all versions of demo/empty game templates shipped. next project on the list is to make a repo that has ags-template-source for each version (with each commit being a diff over the last version).
if you're interested you will get immediate r/w access.

**Comment by ivan-mogilko** [comment]
I'll keep that in mind, but that seem to be mostly useful for analyzing and reverse engineering older games, which is not my priority at the moment. Perhaps it will be more of use to scummvm developers. A while ago someone told me there are people interested in finding much older engine sources (I managed to pass this request to Chris Jones, but he said even if there's anything left - it's on his old computer to which he currently does not have an access).

In regards to this ticket, is the problem with saves resolved?

**Comment by rofl0r** [comment]
> A while ago someone told me there are people interested in finding much older engine sources

indeed, yesterday i had the same idea and wanted to decompile ags 2.7.2, but retdec has removed their online decompiler since they went open-source. the main interest of the older versions is that they run without .net framework, and the ancient wine version i got working can only run plain win32 binaries. having source for such a version would allow to link it against microwindows, libwine, or bgui2 which all emulate win32 api for gui and so have a native editor that could be used cross-platform. 2.6.2 is also interesting because it's the last version before OOP-bloat was introduced, it would allow one to create slim and fast-running games.
if CJ could dig out the sources for those that would be great.

> In regards to this ticket, is the problem with saves resolved?

i didn't test yet because i'm currently on sdl2 branch due to midi support.
one thing that strikes me as a bit awkward is that i have to manually create the 000 file to make it work. wouldn't it easy to deal with this in the engine code ?

**Comment by ivan-mogilko** [comment]
> one thing that strikes me as a bit awkward is that i have to manually create the 000 file to make it work. wouldn't it easy to deal with this in the engine code ?

I don't think save slot 000 is used by ags, are you refering to the problem I mentioned above? If so, it's not engine's issue, it's a bug in the script where it always writes new save in the slot index = number of existing saves + 1, disregarding actual existing slot indexes.

**Comment by rofl0r** [comment]
> are you refering to the problem I mentioned above?

yes, ok - i see now. is it possible you merge the improvements into SDL2 branch ?

actually i noticed that it's custom in AGS development to merge one branch into another, i don't know whether that's the best approach as it probably results in lots of merge errors that need to be fixed manually each time (and ugly merge commits making the history non-linear).
i used to use another approach for similar undertakings: git cherry-pick with specific commits or extracting them to patches by running git format-patch on the branch having the improvements, and then applying them with git am. if a patch produces a merge error the am process can be aborted with git am --abort and the patch manually edited (e.g. by removing hunks that don't apply) or if that's too much work just applying the patch contents manually.

**Comment by ivan-mogilko** [comment]
> is it possible you merge the improvements into SDL2 branch

Yes I could do that.

> actually i noticed that it's custom in AGS development to merge one branch into another, i don't know whether that's the best approach as it probably results in lots of merge errors that need to be fixed manually each time (and ugly merge commits making the history non-linear).

I see what you mean, but there seemed to be at least two pros in such approach: you are not going to miss anything, and you are merging only final state which means you have to solve both code and logical conflicts only once.
Though I hope at some point there will be only ags3 and ags4, and eventually they will just be too separate to merge them into each other, so we'll be forced to use cherry picks and patches anyway to port fixes and whatnot.

**Comment by ivan-mogilko** [comment]
@rofl0r is this question resolved?

Speaking of the save file getting '_' added to the name, the only thing I know when anything is appended to the save file name is when game has a custom save suffix set up in data. But such suffix is always preceded by '.', so it does not match the case either.

**Comment by rofl0r** [comment]
yes, i guess so.
i still didnt get around yet to test larry 2 again, but if there are still issues with savegames i'll open a new ticket.
